### PR TITLE
Added basic Cmake support for Linux

### PR DESCRIPTION
**Pull request by makepath-alex**
Simple usage would be:
```
git clone <repo>
cd <repo>

FC=gfortran
FCFLAGS="-ffree-line-length-none -m64 -std=f2008 -march=native -fbounds-check -fmodule-private -fimplicit-none -finit-real=nan -g -DRTE_USE_CBOOL"
RRTMGP_DATA_VERSION=v1.8.2
FP_MODEL=DP
RTE_BOOL=C
RTE_KERNELS=default
FAILURE_THRESHOLD=7.e-4

cmake -S . -B build \
	-DCMAKE_Fortran_COMPILER=$FC \
	-DCMAKE_Fortran_FLAGS=$FCFLAGS \
	-DRRTMGP_DATA_VERSION=$RRTMGP_DATA_VERSION \
	-DPRECISION=$FP_MODEL \
	-DBOOL_TYPE=$RTE_BOOL \
	-DKERNEL_MODE=$RTE_KERNELS \
	-DENABLE_TESTS=ON \
	-DFAILURE_THRESHOLD=$FAILURE_THRESHOLD
```

This would build the basic libraries in Release configuration

If you want to also build and run the tests, you have to enable the ENABLE_TESTS option, which if OFF by default.
You can do it like so: 
```
FAILURE_THRESHOLD=7.e-4
cmake -S . -B build -DENABLE_TESTS=ON -DFAILURE_THRESHOLD=$FAILURE_THRESHOLD
```
Then build as usual

To run the tests and checks, tun `ctest`, for Example:
```
python3.10 -m venv .venv
source .venv/bin/activate
pip install dask[array] netCDF4 numpy xarray

ctest -v --test-dir build
```
To run the validation-plots, for Example:
```
pip install colorcet matplotlib scipy

cmake --build build --target validation-plots
```

Note that for compiling the tests you need `netcdff` and for running them successfully you would also need the `rrtmgp-data` repo

### CODE REVIEW

**Comment by RobertPincus** [comment]
@makepath-alex This seems to be close, very excting

1. Looking at the CI logs for the "CI in a box" it seems the netCDF installation isn't being found by CMake? 

We could focus a first PR on Linux/Mac and add Windows in a second PR. 

For Windows I'd love CI and  an `environment-windows.yml` file that's used in the CI for installing the needed tools.

**Comment by makepath-alex** [comment]
> @makepath-alex This seems to be close, very excting
> 
> 1. Looking at the CI logs for the "CI in a box" it seems the netCDF installation isn't being found by CMake?
> 
> We could focus a first PR on Linux/Mac and add Windows in a second PR.
> 
> For Windows I'd love CI and an `environment-windows.yml` file that's used in the CI for installing the needed tools.

It is , but it complains it has been compiled with an incompatible compiler. I faced the same issue on Windows and I had to switch to using conda-forge for obtaining both the fortran compiler and the netcdf package.
Will see how I can handle the problem here.

Other than that indeed we seem to be rather close :)

**Comment by RobertPincus** [comment]
@makepath-alex This looks nearly there. Some questions. 

1. Tests are not bring run under the CMake CI? This we want to put back in place (with CTest) before merging. 
2. CI on Linux seems to be working except for a non-essential creation of "validation plots" in containerized CI with `ifort` which I will fix but can be skipped for now. 
3. I will also try to add `cmake` and `git` to the containers, though I've meanwhile uncovered an unhappy change from Intel. 
4. Will Windows including CI be part of this PR?

**Comment by RobertPincus** [comment]
@makepath-alex What's the difference between this PR and #304 ?

**Comment by RobertPincus** [comment]
@makepath-alex
1. I made a new branch `makepath-cmake` for you to merge to. This will let me update the Gitlab CI to CMake before merging to develop. 
2. Looking at the CMake files it seems the tests are all being run in the `build/` directory? If I have that right it'll be easy for me to fix the one error in the containerized CI.

**Comment by RobertPincus** [comment]
@makepath-alex Would it be difficult to add CI for Windows and MacOS? I don't have a good reason for excluding them from the CI in particular. Does CMake make it as easy as specifying different (sets of) compilers and flags for each OS?

**Comment by RobertPincus** [comment]
> > @makepath-alex Would it be difficult to add CI for Windows and MacOS? I don't have a good reason for excluding them from the CI in particular. Does CMake make it as easy as specifying different (sets of) compilers and flags for each OS?
> 
> It might take some time to tweak everything into place for macOS, but it should be relatively quick-in theory. For Windows, it could be more of a challenge. Yes, CMake should handle all the hard work; only the compilers and flags are necessary.

@makepath-alex If this is a matter of a few hours I'd vote for it - that way all the platforms are treated the same, which is draw of using CMake.

File: rte-frontend/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    # FLUX REDUCTION
+    ${CMAKE_CURRENT_SOURCE_DIR}/mo_fluxes.F90
+    # OPTICAL PROPERTIES
+    ${CMAKE_CURRENT_SOURCE_DIR}/mo_optical_props.F90
+    # SOURCE FUNCTIONS
+    ${CMAKE_CURRENT_SOURCE_DIR}/mo_source_functions.F90
+)
+
+add_library(rte STATIC ${SOURCES})
```
[the remark this answers is not in the document]
Comment (makepath-alex) [reply rte-frontend/CMakeLists.txt:16]: In the Make build, we had six libraries: `rte-kernel`, `rrtmgp-kernel`, `rte-frontend`, `rrtmgp-frontend`, `rte`, and `rrtmgp`. From how the Makefile was written, it seemed that `rte-frontend` is equivalent to `rte`, and similarly, `rrtmgp-frontend` is equivalent to `rtetmgp`. I didn’t see any reason to have two sets of seemingly identical libraries. If this is not the case, I’d love to understand the logic behind this setup

**Comment by makepath-alex** [comment]
@RobertPincus From what I see, on the CI, we use ubuntu-22.04. What version should we target for macOS?

**Comment by RobertPincus** [comment]
> @RobertPincus From what I see, on the CI, we use ubuntu-22.04. What version we target for macOS?

Is it greedy to ask for some flavor of MacOS 14 and 15?

File: rte-frontend/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (RobertPincus) [reply rte-frontend/CMakeLists.txt:16]: Good question @makepath-alex. In some permutations we want to test different implementations of the kernels (e.g. in [CUDA](https://github.com/earth-system-radiation/rte-rrtmgp-cuda-kernels)) for which we require the Fortran front-end. In other cases we want to test the Fortran kernels (say with the [C++ front-end](https://github.com/earth-system-radiation/rte-rrtmgp-cpp)). 

As long as we know the separate libraries can be built we don't necessarily need to build them during the CI - but we will want to be able to specify building separate libraries as an option to CMake

[the remark this answers is not in the document]
Comment (RobertPincus) [reply rte-frontend/CMakeLists.txt:16]: (Using *-frontend is fine, I had forgotten the naming)

**Comment by makepath-alex** [comment]
> > @RobertPincus From what I see, on the CI, we use ubuntu-22.04. What version we target for macOS?
> 
> Is it greedy to ask for some flavor of MacOS 14 and 15?

At the moment, I'm looking at [this tool](https://github.com/marketplace/actions/setup-fortran?version=v1.6.0#runner-compatibility) to simplify finding compilers for `Windows` and `macOS`. I managed to pass `continuous-integration` on `macos-13`. The setup for finding and using Fortran compilers on other operating systems (other than Linux) is a bit rough, and I was wondering if there are already existing solutions for this.

**Comment by RobertPincus** [comment]
> > > @RobertPincus From what I see, on the CI, we use ubuntu-22.04. What version we target for macOS?
> > 
> > 
> > Is it greedy to ask for some flavor of MacOS 14 and 15?
> 
> At the moment, I'm looking at [this tool](https://github.com/marketplace/actions/setup-fortran?version=v1.6.0#runner-compatibility) to simplify finding compilers for `Windows` and `macOS`. I managed to pass `continuous-integration` on `macos-13`. The setup for finding and using Fortran compilers on other operating systems (other than Linux) is a bit rough, and I was wondering if there are already existing solutions for this.

@makepath-alex I see the point of testing the Fortran on Windows and MacOS as being to ensure that the kernels don't have some platform-specific bug that will break the conda/Python install. Is managing the compilers any easier if you know you can rely on the conda build tool chain?

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -87,17 +81,21 @@ jobs:
...
-    #
-    # Compare the results
-    #
-    - name: Compare the results
-      run: make -j4 check
+        cmake -S . -B build \
+            -DCMAKE_Fortran_COMPILER=$FC \
+            -DCMAKE_Fortran_FLAGS="$FCFLAGS" \
+            -DRRTMGP_DATA_VERSION=$RRTMGP_DATA_VERSION \
+            -DFP_MODEL=$FP_MODEL \
```
Comment (RobertPincus) [inline .github/workflows/continuous-integration.yml:91]: @makepath-alex Can I ask why you moved $FP_MODEL to the CMake configuration while leaving $RTE_USE_CBOOL as a compiler flag? The two choices are implemented in the same way (with \#IFDEFS in the code).

Comment (makepath-alex) [reply .github/workflows/continuous-integration.yml:91]: Missed one :D

File: rrtmgp-kernels/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,27 @@
+set(LOCAL_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+
+if (KERNEL_MODE STREQUAL "accel")
+    set(LOCAL_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/accel)
+elseif(KERNEL_MODE STREQUAL "extern")
+    set(LOCAL_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/api)
+endif()
+
+if (KERNEL_MODE STREQUAL "accel")
```
Comment (RobertPincus) [inline rrtmgp-kernels/CMakeLists.txt:9]: @makepath-alex Shouldn't this be `LOCAL_SOURCE_DIR` ?

File: rte-frontend/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (RobertPincus) [reply rte-frontend/CMakeLists.txt:16]: @makepath-alex Does it make sense to build the kernel and front-end libraries separately and then combine them to make `librte` and `librrtmgp`? I'd like to make this explicit.

**Comment by RobertPincus** [comment]
@skosukhin I'm curious to hear any feedback you might have.

File: rte-frontend/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (makepath-alex) [reply rte-frontend/CMakeLists.txt:16]: Let's discuss this over the call in the evening as I am a little confused and I suspect there's some terminology included here that causes the confision. We'll clear it up.

**Pull request merged**